File: md_studio/utils/toc.py
import re
from typing import List, Dict
from dataclasses import dataclass

@dataclass
class TocItem:
    id: str
    title: str
    level: int

def generate_heading_id(text: str) -> str:
    heading_id = text.lower()
    heading_id = re.sub(r'[^a-z0-9\s-]', '', heading_id)
    heading_id = re.sub(r'\s+', '-', heading_id)
    heading_id = re.sub(r'-+', '-', heading_id)
    return re.sub(r'^-|-$', '', heading_id)

def extract_table_of_contents(markdown: str) -> List[TocItem]:
    heading_regex = re.compile(r'^(#{2,3})\s+(.+)$', re.MULTILINE)
    toc: List[TocItem] = []
    id_counts: Dict[str, int] = {}
    
    for match in heading_regex.finditer(markdown):
        level = len(match.group(1))
        title = match.group(2).strip()
        heading_id = generate_heading_id(title)
        
        count = id_counts.get(heading_id, 0)
        if count > 0:
            heading_id = f"{heading_id}-{count}"
        id_counts[generate_heading_id(title)] = count + 1
        
        toc.append(TocItem(id=heading_id, title=title, level=level))
    
    return toc

class HeadingIdGenerator:
    def __init__(self):
        self.id_counts: Dict[str, int] = {}
    
    def generate(self, text: str) -> str:
        base_id = generate_heading_id(text)
        count = self.id_counts.get(base_id, 0)
        
        heading_id = base_id
        if count > 0:
            heading_id = f"{base_id}-{count}"
        
        self.id_counts[base_id] = count + 1
        return heading_id
    
    def reset(self) -> None:
        self.id_counts.clear()

File: ui/lib/toc.ts
export type TocItem = {
  id: string;
  title: string;
  level: number;
};

/**
 * Generate a URL-safe ID from heading text
 * Matches the behavior of most markdown renderers
 */
export function generateHeadingId(text: string): string {
  return text
    .toLowerCase()
    .replace(/[^a-z0-9\s-]/g, "") // Remove special chars
    .replace(/\s+/g, "-") // Replace spaces with hyphens
    .replace(/-+/g, "-") // Collapse multiple hyphens
    .replace(/^-|-$/g, ""); // Remove leading/trailing hyphens
}

/**
 * Extract table of contents from markdown content
 * Generates IDs from heading text for linking
 */
export function extractTableOfContents(markdown: string): TocItem[] {
  const headingRegex = /^(#{2,3})\s+(.+)$/gm;
  const toc: TocItem[] = [];
  const idCounts = new Map<string, number>();
  let match: RegExpExecArray | null;

  while ((match = headingRegex.exec(markdown)) !== null) {
    const level = match[1].length; // 2 for ##, 3 for ###
    const title = match[2].trim();
    let id = generateHeadingId(title);

    // Handle duplicate IDs by appending a counter
    const count = idCounts.get(id) ?? 0;
    if (count > 0) {
      id = `${id}-${count}`;
    }
    idCounts.set(generateHeadingId(title), count + 1);

    toc.push({ id, title, level });
  }

  return toc;
}

/**
 * Helper class to track heading IDs and ensure uniqueness
 * Use this when rendering markdown to match TOC IDs
 */
export class HeadingIdGenerator {
  private idCounts = new Map<string, number>();

  generate(text: string): string {
    const baseId = generateHeadingId(text);
    const count = this.idCounts.get(baseId) ?? 0;
    
    let id = baseId;
    if (count > 0) {
      id = `${baseId}-${count}`;
    }
    
    this.idCounts.set(baseId, count + 1);
    return id;
  }

  reset(): void {
    this.idCounts.clear();
  }
}
